Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs

```diff
@@ -0,0 +1,40 @@
+module SharedLogic.FarePolicy where
+
+import Data.List (nub)
+import Domain.Types.FarePolicy
+import Domain.Types.Merchant
+import Domain.Types.Vehicle
+import Kernel.Prelude
+import Kernel.Storage.Esqueleto.Config
+import Kernel.Types.Id
+import Kernel.Utils.Common
+import qualified Lib.Types.SpecialLocationLink as SpecialLocationLink
+import qualified SharedLogic.SpecialLocation as SpecialLocation
+import Storage.CachedQueries.CacheConfig
+import qualified Storage.CachedQueries.FarePolicy as QFP
+import Tools.Maps
+
+getFlowType :: (MonadFlow m, MonadReader r m, HasEsqEnv m r, CacheFlow m r) => Id Merchant -> LatLong -> LatLong -> m SpecialLocationLink.FlowType
+getFlowType merchantId fromlocaton toLocation = do
+  mbSpecialLocationLinks <- SpecialLocation.getSpecialLocationLinkForVariants merchantId.getId fromlocaton toLocation
+  case mbSpecialLocationLinks of
+    Just specialLocationLinks -> do
+      let mbFlowType = listToMaybe $ nub $ (.flow) <$> specialLocationLinks
+      return $ fromMaybe SpecialLocationLink.NORMAL mbFlowType
+    Nothing -> return SpecialLocationLink.NORMAL
+
+getFarePolicyForVariant :: (MonadFlow m, MonadReader r m, HasEsqEnv m r, CacheFlow m r) => Id Merchant -> LatLong -> LatLong -> Variant -> m (Maybe FarePolicy)
+getFarePolicyForVariant merchantId fromlocaton toLocation vehVariant = do
+  mbSpecialLocationLink <- SpecialLocation.getSpecialLocationLinkForVariant merchantId.getId fromlocaton toLocation vehVariant
+  case mbSpecialLocationLink of
+    Just specialLocationLink -> QFP.findById (Id specialLocationLink.farePolicyId)
```

Comment:
```
1. rename FarePolicyDetails to FarePolicyTypeDetails
2. create new 
data FarePolicyTrait = NormalTrait FarePolicyTypeDetails | SpecialZonesTrait FarePolicySpecialZoneTrait
data FarePolicySpecialZoneTrait = FarePolicySpecialZoneTrait {
  -- there you implement all your task fields and necessaries like pickupPriority and other stuff, for each pickupPriority you'll have its own FarePolicyTypeDetails
}
data FarePolicyTraitsListItem = FarePolicyTraitListItem {
   farePolicyTrait :: FarePolicyTrait,
   traitPriority :: Int
}
3. replace FarePolicy.farePolicyDetails with FarePolicy.farePolicyTraitsList (NonEmpty, always have item with priority 0, which is most likely just NormalTrait)

Now when you passing ride info to count fareParams, you have to process this list from highest priority number to 0 (which is default case, which can handle just everything) and choose Trait with highest priority (currently we will have only 0 and 1 priorities, where 1 for FarePolicySpecialZoneTrait, but potentially we can create more easy enough) and count FareParams based on choosen trait. 
```
That's how i proposed to implement this and i as far as i understand your way is far from proposed approach. Did you find some issues with this? Or it was just too hard to implement?
@hkmangla I'd like to implement it that way, do we have some time before merging this, so i could refactor this PR?

Reply:
@DonDokidon Yes, but with this approach for 72 Special Locations will have 72 different fare policies even though if they are same, it would get hard to maintain right ?

Reply:
No? You will have only one FarePolicy, which will contain only one `FPSpecZoneTrait`, which will have 72 different `FPSpecZoneTraitZoneCosts` (idk why it's 72 though, you do care only on highest priority, so you'll need only 10 of ZoneCosts entities, 5 for highest pickup. 5 for highest drops). And even if you wanna have one single `FPSpecZoneTraitZoneCosts` for all 10 types of pickup/drop priority, you can make that with this approach, but IMO THAT will be harder to maintain. It will be easy to add directly from DB, by creating one `FPSpecZoneTraitZoneCosts` and assigning that to all 10 pickup/drops, using id ref. But how will you do it, when you will set it via dashboard? Will you check each pickup/drop if their settings are exactly the same and will create one single `FPSpecZoneTraitZoneCosts` entity for them in that case?